### PR TITLE
ddl: forbid add column with sequence as default value

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Problem Summary: forbid add column with sequence as default value
```
Background: the new added column in TiDB won't fill the known rows with specific
sequence next value immediately. Every time TiDB select the data from storage, kvDB
will fill the originDefaultValue to these incomplete rows (but not store).

In sequence case, every time filling these rows, kvDB should eval the sequence
expr for len(incomplete rows) times, and combine these row data together. That
means the select result is not always the same.

However, the altered column with sequence as it's default value can work well.
Because this column has already been added before the alter action, which also
means originDefaultValue should be something but nil, so the back filling in kvDB
can work well.

The new altered sequence default value for this column only take effect on the
subsequent inserted rows.

So under current situation, TiDB will
[1]: forbid the new added column has sequence as it's default value.
[2]: allow the altered column with sequence as default value.
```
### What is changed and how it works?

How it Works: forbid add column with sequence as default value

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note 
- ddl: forbid add column with sequence as default value

### CODE REVIEW

**Comment by AilinKid** [comment]
should we cherry-pick this to release-4.0?

File: ddl/sequence_test.go [hunk 1]
```diff
@@ -918,3 +925,50 @@ func (s *testSequenceSuite) TestSequenceFunctionPrivilege(c *C) {
...
+// means originDefaultValue should be something but nil, so the back filling in kvDB
+// can work well.
+//
+// The new altered sequence default value for this column only take effect on the
+// subsequent inserted rows.
+//
+// So under current situation, TiDB will
+// [1]: forbid the new added column has sequence as it's default value.
+// [2]: allow the altered column with sequence as default value.
+func (s *testSequenceSuite) TestSequenceDefaultLogic(c *C) {
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply ddl/sequence_test.go:948]: only `nextval` can be parsed through as default value in the parser.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2964,6 +2977,14 @@ func checkModifyTypes(origin *types.FieldType, to *types.FieldType, needRewriteC
...
 }
 
+func checkDefaultValueIsExpr(option *ast.ColumnOption) (bool, error) {
+	_, isSeqExpr, err := tryToGetSequenceDefaultValue(option)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return isSeqExpr, nil
+}
+
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply ddl/ddl_api.go:2987]: good

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -2983,6 +2996,11 @@ func checkModifyTypes(origin *types.FieldType, to *types.FieldType, needRewriteC
 	return errors.Trace(err)
 }
 
+func checkDefaultValueIsExpr(option *ast.ColumnOption) (bool, error) {
+	_, isSeqExpr, err := tryToGetSequenceDefaultValue(option)
+	return isSeqExpr, errors.Trace(err)
+}
+
```
Comment (djshow832) [inline ddl/ddl_api.go:3003]: Why not remove this function, and check it directly in `checkAndCreateNewColumn`?

Comment (AilinKid) [reply ddl/ddl_api.go:3003]: maybe the function name is not suitable in here

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2432,6 +2432,19 @@ func checkAndCreateNewColumn(ctx sessionctx.Context, ti ast.Ident, schema *model
 				return nil, errors.Trace(err)
 			}
 		}
+		// Specially, since sequence has been supported, if a new added column has a
```
Comment (zimulala) [inline ddl/ddl_api.go:2435]: s/new/newly

File: ddl/sequence_test.go [hunk 5]
```diff
@@ -918,3 +925,50 @@ func (s *testSequenceSuite) TestSequenceFunctionPrivilege(c *C) {
 	s.tk.MustExec("drop sequence seq")
 	s.tk.MustExec("drop user myuser@localhost")
 }
+
+// Background: the new added column in TiDB won't fill the known rows with specific
```
Comment (zimulala) [inline ddl/sequence_test.go:929]: Ditto.

**Comment by AilinKid** [comment]
@djshow832 addressed PTAL

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -2432,6 +2432,19 @@ func checkAndCreateNewColumn(ctx sessionctx.Context, ti ast.Ident, schema *model
 				return nil, errors.Trace(err)
 			}
 		}
+		// Specially, since sequence has been supported, if a newly added column has a
+		// sequence nextval function as it's default value option, it won't fill the
+		// known rows with specific sequence next value under current add column logic.
+		// More explanation can refer: TestSequenceDefaultLogic's comment in sequence_test.go
+		if option.Tp == ast.ColumnOptionDefaultValue {
+			_, isSeqExpr, err := tryToGetSequenceDefaultValue(option)
```
Comment (djshow832) [inline ddl/ddl_api.go:2440]: I think `option.Expr.(*ast.FuncCallExpr)` is enough.

**Pull request merged**